### PR TITLE
Test Iceberg split pruning when filtering on partition column

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

**Comment by findepi** [comment]
CI https://github.com/trinodb/trino/issues/8432 (`trino-tests`)

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 1]
```diff
@@ -1995,6 +2000,72 @@ public void testIncorrectIcebergFileSizes()
...
+                "SELECT * FROM " + tableName);
+
+        ResultWithQueryId<MaterializedResult> selectSinglePartitionResult = getDistributedQueryRunner().executeWithQueryId(
+                getSession(),
+                "SELECT * FROM " + tableName + " WHERE regionkey = 3");
+
+        ResultWithQueryId<MaterializedResult> selectTwoPartitionsResult = getDistributedQueryRunner().executeWithQueryId(getSession(),
+                "SELECT * FROM " + tableName + " WHERE regionkey < 2");
+
+        verifySplitCount(selectAllPartitionsResult.getQueryId(), 5);
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2026]: annotate the test as likely flaky due to #5172

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 2]
```diff
@@ -1995,6 +2000,72 @@ public void testIncorrectIcebergFileSizes()
...
+        try {
+            OperatorStats operatorStats = getOperatorStats(queryId);
+            assertThat(operatorStats.getTotalDrivers()).isEqualTo(expectedCount);
+        }
+        catch (NoSuchElementException e) {
+            throw new RuntimeException("Couldn't find operator summary, probably due to query statistic collection error", e);
+        }
+    }
+
+    private void verifyNoSplits(QueryId queryId)
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2044]: unused

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 3]
```diff
@@ -1995,6 +2000,72 @@ public void testIncorrectIcebergFileSizes()
...
+        ResultWithQueryId<MaterializedResult> selectSinglePartitionResult = getDistributedQueryRunner().executeWithQueryId(
+                getSession(),
+                "SELECT * FROM " + tableName + " WHERE regionkey = 3");
+
+        ResultWithQueryId<MaterializedResult> selectTwoPartitionsResult = getDistributedQueryRunner().executeWithQueryId(getSession(),
+                "SELECT * FROM " + tableName + " WHERE regionkey < 2");
+
+        verifySplitCount(selectAllPartitionsResult.getQueryId(), 5);
+        verifySplitCount(selectSinglePartitionResult.getQueryId(), 1);
+        verifySplitCount(selectTwoPartitionsResult.getQueryId(), 2);
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2028]: add a test case

- `regionkey < 0` -- everything is filtered out
- `regionkey % 5 = 3` -- partitions should still be pruned by means of `Contraint#predicate`. We should have a follow up if they are not

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 4]
```diff
@@ -1995,6 +2000,72 @@ public void testIncorrectIcebergFileSizes()
...
+
+        dropTable(tableName);
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2030]: that's exuberant. we don't need to test that drop table drops a table
`assertQuery("DROP TABLE " + tableName)` would suffice

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 5]
```diff
@@ -1995,6 +2000,72 @@ public void testIncorrectIcebergFileSizes()
...
     }
 
+    @Test
+    public void testSplitPruningForFilterOnPartitionColumn()
+    {
+        String tableName = "nation_partitioned_pruning";
+
+        // disable writes redistribution to have predicatable number of files written per partition (one).
+        Session noRedistributeWrites = Session.builder(getSession())
+                .setSystemProperty("redistribute_writes", "false")
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2010]: Verify that there exactly one file.
you can do so with select from `$files`.

**Review by findepi — APPROVED** [review]

**Comment by losipiuk** [comment]
AC. Not sure if I put proper `match` for `@Flaky` annotation as it is always passing on my machine.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 6]
```diff
@@ -1995,6 +2001,74 @@ public void testIncorrectIcebergFileSizes()
...
+                .build();
+
+        assertUpdate(noRedistributeWrites, "CREATE TABLE " + tableName + " WITH (partitioning = ARRAY['regionkey']) AS SELECT * FROM nation", 25);
+
+        // sanity check that table contains exactly 5 files
+        assertThat(query("SELECT count(*) FROM \"" + tableName + "$files\"")).matches("VALUES CAST(5 AS BIGINT)");
+
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey < 0", 0);
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2022]: Maybe sth like `WHERE regionkey > 2 AND regionkey <= 4`?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 7]
```diff
@@ -1995,6 +2001,74 @@ public void testIncorrectIcebergFileSizes()
...
+        verifySplitCount("SELECT * FROM " + tableName, 5);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey = 3", 1);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey < 2", 2);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey < 0", 0);
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2026]: Yeah - artifact from debugging. removing

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java (hunk 6 above)
Comment (losipiuk) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2022]: Added sth

**Comment by losipiuk** [comment]
rebased + AC. 
I also added `assertEventually` due to https://github.com/trinodb/trino/issues/5172

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 8]
```diff
@@ -2030,6 +2039,78 @@ public void testIncorrectIcebergFileSizes()
...
+        assertUpdate(noRedistributeWrites, "CREATE TABLE " + tableName + " WITH (partitioning = ARRAY['regionkey']) AS SELECT * FROM nation", 25);
+
+        // sanity check that table contains exactly 5 files
+        assertThat(query("SELECT count(*) FROM \"" + tableName + "$files\"")).matches("VALUES CAST(5 AS BIGINT)");
+
+        verifySplitCount("SELECT * FROM " + tableName, 5);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey = 3", 1);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey < 2", 2);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey < 0", 0);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey >1 AND regionkey < 4", 2);
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2064]: Spacing Nit 
```suggestion
        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey > 1 AND regionkey < 4", 2);
```

**Review by alexjo2144 — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 9]
```diff
@@ -2030,6 +2039,78 @@ public void testIncorrectIcebergFileSizes()
...
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey % 5 = 3", 5);
+
+        assertUpdate("DROP TABLE " + tableName);
+    }
+
+    private void verifySplitCount(String query, int expectedSplitCount)
+    {
+        // using assertEventually here as the operator stats mechanism is known to be best-effort and some late-stage stats are sometimes not recorded
+        // (see https://github.com/trinodb/trino/issues/5172)
+        assertEventually(new Duration(10, TimeUnit.SECONDS), () -> {
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2076]: Shouldn't the `executeWithQueryId` be outside of the eventually loop?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 10]
```diff
@@ -2030,6 +2039,78 @@ public void testIncorrectIcebergFileSizes()
...
 
+    @Test
+    @Flaky(issue = "https://github.com/trinodb/trino/issues/5172", match = "Couldn't find operator summary, probably due to query statistic collection error")
+    public void testSplitPruningForFilterOnPartitionColumn()
+    {
+        String tableName = "nation_partitioned_pruning";
+
+        assertUpdate("DROP TABLE IF EXISTS " + tableName);
+
+        // disable writes redistribution to have predicatable number of files written per partition (one).
```
Comment (phd3) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2050]: did you mean `predicatable` or `predictable`?

**Review by phd3 — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java (hunk 10 above)
Comment (losipiuk) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2050]: predictable

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java (hunk 9 above)
Comment (losipiuk) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:2076]: IDK to be honest. IIRC correctly we observed that sometimes statistics were never recorded for completed query; even after some time. So having it outside is safer if chance of success is big enough. 
The most safe approach would be to have longer assertEventually outside and shorter inside, but lets not do that unless we have to.

**Pull request merged**